components/RenderErrorBoundary.test.tsx

```diff
@@ -0,0 +1,131 @@
+            </>
+          )}
+        </>
+      )
+    })
+
+    const { container, getByText } = render(<RenderErrorBoundary>{mockRenderProp}</RenderErrorBoundary>)
+
+    const error = expect.any(Error)
+    const info = expect.any(Object)
+    const clear = expect.any(Function)
+
+    expect(mockRenderProp).toHaveBeenCalledTimes(2)
+    expect(mockRenderProp).toHaveBeenNthCalledWith(1, { renderError: null })
+    expect(mockRenderProp).toHaveBeenNthCalledWith(2, { renderError: { clear, error, info } })
+    expect(container).toHaveTextContent('Oh no! Not again!')
+
+    fireEvent.click(getByText('Clear'))
+
+    expect(mockRenderProp).toHaveBeenCalledTimes(4)
+    expect(mockRenderProp).toHaveBeenNthCalledWith(3, { renderError: null })
+    expect(mockRenderProp).toHaveBeenNthCalledWith(4, { renderError: { clear, error, info } })
+  } finally {
+    ;(console.error as jest.Mock).mockRestore()
+  }
+})
+
+test('RenderErrorBoundary called with `onClear` or `onError` passes render prop an argument object with a `renderError` property with an error, info, and a clear function when render prop errors', () => {
+  try {
+    // Temporarily turn off the error console.
+    jest.spyOn(console, 'error')
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    ;(console.error as jest.Mock).mockImplementation(() => {})
+
+    const mockRenderProp = jest.fn(({ renderError }) => {
+      return (
+        <>
+          {renderError === null ? (
+            <BadComponent />
+          ) : (
+            <>
+              <p>Oh no! Not again!</p>
+              <button onClick={renderError.clear}>Clear</button>
+            </>
+          )}
+        </>
+      )
+    })
+
+    const onClear = jest.fn()
+    const onError = jest.fn()
+    const { container, getByText } = render(
+      <RenderErrorBoundary onClear={onClear} onError={onError}>
+        {mockRenderProp}
+      </RenderErrorBoundary>,
+    )
+
+    const error = expect.any(Error)
+    const info = expect.any(Object)
+    const clear = expect.any(Function)
+
+    expect(onClear).toHaveBeenCalledTimes(0)
+
+    expect(onError).toHaveBeenCalledTimes(1)
+    expect(onError).toHaveBeenNthCalledWith(1, { clear, error, info })
+
+    expect(mockRenderProp).toHaveBeenCalledTimes(2)
+    expect(mockRenderProp).toHaveBeenNthCalledWith(1, { renderError: null })
+    expect(mockRenderProp).toHaveBeenNthCalledWith(2, { renderError: { clear, error, info } })
+    expect(container).toHaveTextContent('Oh no! Not again!')
+
+    fireEvent.click(getByText('Clear'))
+
+    expect(onClear).toHaveBeenCalledTimes(1)
+
+    expect(onError).toHaveBeenCalledTimes(2)
+    expect(onError).toHaveBeenNthCalledWith(2, { clear, error, info })
+
+    expect(mockRenderProp).toHaveBeenCalledTimes(4)
+    expect(mockRenderProp).toHaveBeenNthCalledWith(3, { renderError: null })
+    expect(mockRenderProp).toHaveBeenNthCalledWith(4, { renderError: { clear, error, info } })
```

Comment:
It's a bit difficult to reason how you got these expects and it might be worth adding a comment above each of them to explain what it means when they fail.

```suggestion
    // called after attempt to render bad component, now with an error.
    expect(mockRenderProp).toHaveBeenNthCalledWith(4, { renderError: { clear, error, info } })
```

This way someone debugging or extending the error boundary doesn't need to dissect the entire component to understand why it failed. If this specific case failed, they can now realize their changes made it so that the error wasn't handled correctly.

Reply:
Added some comments.